Fix InvariantGlobalization=false in a trimmed app

eerhardt opened the pull request:
Move the LoadICU logic into a static ctor, so it runs early in the process, but not as part of querying the GlobalizationMode.Invariant property. This allows for LoadICU to still be invoked, even when the app is trimmed and GlobalizationMode.Invariant is hard-coded `false` by the linker.

While I was changing this code, I also removed the workaround in Browser builds for swallowing errors being returned from LoadICU.

Fix #49073
Fix #49391

eerhardt committed:
Fix InvariantGlobalization=false in a trimmed app.

Move the LoadICU logic into a static ctor, so it runs early in the process, but not as part of querying the GlobalizationMode.Invariant property. This allows for LoadICU to still be invoked, even when the app is trimmed and GlobalizationMode.Invariant is hard-coded by the linker.

While I was changing this code, I also removed the workaround in Browser builds for swallowing errors being returned from LoadICU.

Fix #49073
Fix #49391

eerhardt committed:
Add trimming tests for InvariantGlobalization

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Unix.cs:
```diff
@@ -7,42 +7,37 @@ internal static partial class GlobalizationMode
+            /// Load ICU (when not in Invariant mode) in a static cctor to ensure it is loaded early in the process.
+            /// Other places, e.g. CompareInfo.GetSortKey, rely on ICU already being loaded before they are called.
+            /// </summary>
+            static Settings()
```
>static [](http://example.com/codeflow?start=12&length=6)

private?

eerhardt replied:
It’s a static constructor. You can’t put accessibility modifiers on them.

tarekgh commented on src/libraries/System.Runtime/tests/TrimmingTests/InvariantGlobalizationFalse.cs:
```diff
@@ -0,0 +1,26 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
>// [](http://example.com/codeflow?start=0&length=2)

do you think we need to have any test when enabling NLS mode on Windows?

eerhardt replied:
Do we have a feature switch for UseNls? Does that property participate in trimming at all?

tarekgh replied:
> Do we have a feature switch for UseNls? 

Yes, https://github.com/dotnet/runtime/blob/f881cf3839d44fc356dc3a4dd1c1b7b19cc82b66/src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Windows.cs#L12.

> Does that property participate in trimming at all?

I don't think so. I am trying to ensure Invariant and NLS behaviors modes not affected. I was just asking if we need to do something more here. We already have tests specific for NLS which should be covering that. If you think nothing more need to be done, you can ignore this comment.

eerhardt replied:
I think the current tests will suffice.

tarekgh reviewed (approved):
Added a minor comment and question, LGTM otherwise.

lewing commented on src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Unix.cs:
```diff
@@ -7,42 +7,37 @@ internal static partial class GlobalizationMode
+                        if (loaded == 0)
+                        {
+                            // This can't go into resources, because a resource lookup requires globalization, which requires ICU
+                            string message = "Couldn't find a valid ICU package installed on the system. " +
+                                             "Please install libicu using your package manager and try again. " +
+                                             "Alternatively you can set the configuration flag System.Globalization.Invariant to true if you want to run with no globalization support. " +
+                                             "Please see https://aka.ms/dotnet-missing-libicu for more information.";
```
This message doesn't make sense for browser does it?  And the link also doesn't offer any solution.

marek-safar replied:
The message does not make sense for other workloads too. What do you think developers will do if they see it when running on Android?

eerhardt replied:
Do you have suggestions for what it should say when ICU comes with the app? Basically it means we have a bug in our code and the app was packaged incorrectly, so what should a user do in this case?

marek-safar replied:
I don't think we need to pretend we know what is wrong. A simple error message that file xyz could not be loaded would work

lewing replied:
or since it is not always a file  "Unable to load required ICU Globalization data. Please see https://aka.ms/dotnet-missing-libicu for more information"

marek-safar commented on src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml:
```diff
@@ -6,7 +6,7 @@
       <method signature="System.Boolean IsEnabled(System.Diagnostics.Tracing.EventLevel,System.Diagnostics.Tracing.EventKeywords,System.Diagnostics.Tracing.EventChannel)" body="stub" value="false" />
     </type>
-    <type fullname="System.Globalization.GlobalizationMode">
+    <type fullname="System.Globalization.GlobalizationMode/Settings">
```
This effectively removes `Invariant = true` dependencies trimming. I think it'd be better to substitute at the current level.

eerhardt replied:
> This effectively removes Invariant = true dependencies trimming.

Why?

eerhardt replied:
I tested it and the dead branches still get trimmed. Why do you think this removes that?

Also, you can’t substitute this at the higher level because then the Settings class wouldn’t be touched and it’s cctor won’t be run.

And there were concerns in the issue about putting the cctor on GlobalizationMode itself because calling UseNls would force it to run. So splitting the cctor off in a separate class was proposed.

marek-safar replied:
> Also, you can’t substitute this at the higher level because then the Settings class wouldn’t be touched and it’s cctor won’t be run.

I think that's what we want for `Invariant = true`. I think Settings for that mode is redundant.

eerhardt replied:
I'm not sure how to resolve this comment, plus address all the concerns raised in https://github.com/dotnet/runtime/issues/49073#issuecomment-794550693. Can you elaborate on what you want here?

Do you want me to split the ILLink.Substitutions.xml apart so `Invariant = true` and `Invariant = false` look different? For example:

```xml
    <type fullname="System.Globalization.GlobalizationMode">
      <method signature="System.Boolean get_Invariant()" body="stub" value="true" feature="System.Globalization.Invariant" featurevalue="true" />
    </type>
    <type fullname="System.Globalization.GlobalizationMode/Settings">
      <method signature="System.Boolean get_Invariant()" body="stub" value="false" feature="System.Globalization.Invariant" featurevalue="false" />
    </type>
```

That doesn't seem like an ideal situation IMO.

I don't think your original concern is a problem. Can you explain what are the disadvantages to the current approach I am using?

marek-safar commented on src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Unix.cs:
```diff
@@ -7,42 +7,37 @@ internal static partial class GlobalizationMode
+            static Settings()
             {
-                if (TryGetAppLocalIcuSwitchValue(out string? icuSuffixAndVersion))
+                if (!Invariant)
```
I'm not a big fan of mixing static ctor with static fields initializers. If you keep the substitution at higher level you could better control when GetInvariantSwitchValue is called

eerhardt replied:
What's the concern here? The static field initializers will run at the beginning of the static cctor, just like if you had instance field initializers and a ctor.

eerhardt committed:
Update the LoadICU message for mobile workloads.

eerhardt committed:
Respond to PR feedback.

- Split the substitutions of GlobalizationMode.Invariant between true and false
- Add a trimming test that ensures Invariant=true trims everything it is supposed to

eerhardt commented:
@marek-safar - I believe I have addressed your concerns with the latest changes. PTAL.

marek-safar commented on src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Unix.cs:
```diff
@@ -7,42 +7,48 @@ internal static partial class GlobalizationMode
             {
-                if (TryGetAppLocalIcuSwitchValue(out string? icuSuffixAndVersion))
+                // These strings can't go into resources, because a resource lookup requires globalization, which requires ICU
+                if (OperatingSystem.IsBrowser() || OperatingSystem.IsIOS() || OperatingSystem.IsAndroid())
```
I'd go with the simple message everywhere. Otherwise, you would also need to check for tvOS, MacCatalyst, macOS.

eerhardt replied:
I'll add the check for `tvOS` and `MacCatalyst`. My concern with using the "simple" message is that for "desktop" scenarios like Linux and macOS, it would be less informative than it is today.

marek-safar replied:
I don't think macOS users will know what to do with the message and in the case of macOS UI apps it makes even less sense if the user somehow manages to see it.

eerhardt replied:
I believe it does make sense for macOS users. We require they have ICU on the machine. We don't bring ICU with .NET on macOS.

See https://docs.microsoft.com/en-us/dotnet/standard/globalization-localization/globalization-icu#macos-behavior.

eerhardt replied:
@marek-safar - I'm going to merge this for now. If we decide to do something different here for macOS, we can open a new issue/PR for it. But, as-is, this change is keeping the current/existing behavior for macOS.

eerhardt committed:
Add checks for all mobile platforms
